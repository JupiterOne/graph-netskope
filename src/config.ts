import {
  IntegrationExecutionContext,
  IntegrationValidationError,
  IntegrationInstanceConfigFieldMap,
  IntegrationInstanceConfig,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from './client';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  tenantName: {
    type: 'string',
  },
  apiV1Token: {
    type: 'string',
    mask: true,
  },
};

export interface IntegrationConfig extends IntegrationInstanceConfig {
  tenantName: string;
  /**
   * Server generated API V1 Token for request authentication
   */
  apiV1Token: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { config } = context.instance;

  if (!config.tenantName || !config.apiV1Token) {
    throw new IntegrationValidationError(
      'Config requires all of the following {tenantName, apiV1Token}. Please generate token or request access from server admin.',
    );
  }

  const apiClient = createAPIClient(config, context.logger);
  await apiClient.verifyAuthentication();
}
