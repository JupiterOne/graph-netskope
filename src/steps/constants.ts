import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const TENANT_ENTITY_KEY = 'entity:tenant';

export const IntegrationSteps = {
  TENANT: 'fetch-tenant',
  APP_INSTANCES: 'fetch-app-instances',
  DEVICES: 'fetch-devices',
  BUILD_DEVICE_USER_RELATIONSHIPS: 'build-device-has-user-relationships',
  FETCH_AND_BUILD_USER_CONFIGURATION:
    'fetch-and-build-user-has-configuration-relationships',
};

export const Entities: Record<
  'TENANT' | 'DEVICE' | 'USER' | 'USER_CONFIGURATION' | 'APP_INSTANCE',
  StepEntityMetadata
> = {
  TENANT: {
    resourceName: 'Tenant',
    _type: 'netskope_tenant',
    _class: ['Organization'],
    schema: {
      properties: {
        id: { type: 'string' },
        name: { type: 'string' },
      },
      required: ['id'],
    },
  },
  DEVICE: {
    resourceName: 'Device',
    _type: 'netskope_device',
    _class: ['Device'],
    schema: {
      properties: {
        category: { type: 'string' },
        make: { type: 'string' },
        model: { type: 'string' },
        serial: { type: 'string' },
        deviceId: { type: 'string' },
        name: { type: 'string' },
        id: { type: 'string' },
        clientInstallTime: { type: 'number' },
        clientVersion: { type: 'string' },
        os: { type: 'string' },
        osVersion: { type: 'string' },
        lastEvent: { type: 'string' },
        lastEventStatus: { type: 'string' },
        lastEventNpaStatus: { type: 'string' },
        lastEventActor: { type: 'string' },
        lastEventOccurredOn: { type: 'number' },
      },
      required: ['id'],
    },
  },
  USER: {
    resourceName: 'User',
    _type: 'netskope_user',
    _class: ['User'],
    schema: {
      properties: {
        id: { type: 'string' },
        username: { type: 'string' },
        name: { type: 'string' },
        email: { type: 'string' },
        active: { type: 'boolean' },
        userAddedTime: { type: 'number' },
        userSource: { type: 'string' },
        lastEvent: { type: 'string' },
        lastEventStatus: { type: 'string' },
        lastEventNpaStatus: { type: 'string' },
        lastEventActor: { type: 'string' },
        lastEventOccurredOn: { type: 'number' },
      },
      required: ['username', 'active', 'id', 'email'],
    },
  },
  USER_CONFIGURATION: {
    resourceName: 'User Configuration',
    _type: 'netskope_user_configuration',
    _class: ['Configuration'],
    schema: {
      properties: {
        name: { type: 'string' },
        email: { type: 'string' },
        addonCheckerHost: { type: 'string' },
        addonCheckerResponseCode: { type: 'string' },
        addonManagerHost: { type: 'string' },
        orgName: { type: 'string' },
        SFCheckerHost: { type: 'string' },
        SFCheckerIP: { type: 'string' },
      },
      required: ['email'],
    },
  },
  APP_INSTANCE: {
    resourceName: 'App Instance',
    _type: 'netskope_app_instance',
    _class: ['Application'],
    schema: {
      properties: {
        id: { type: 'string' },
        name: { type: 'string' },
        app: { type: 'string' },
        type: { type: 'string' },
        lastModified: { type: 'string' },
      },
      required: ['id'],
    },
  },
};

export const Relationships: Record<
  | 'TENANT_HAS_DEVICE'
  | 'DEVICE_HAS_USER'
  | 'USER_HAS_USER_CONFIGURATION'
  | 'TENANT_HAS_APP_INSTANCE',
  StepRelationshipMetadata
> = {
  TENANT_HAS_DEVICE: {
    _type: 'netskope_tenant_has_device',
    sourceType: Entities.TENANT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.DEVICE._type,
  },
  DEVICE_HAS_USER: {
    _type: 'netskope_device_has_user',
    sourceType: Entities.DEVICE._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.USER._type,
  },
  USER_HAS_USER_CONFIGURATION: {
    _type: 'netskope_user_has_configuration',
    sourceType: Entities.USER._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.USER_CONFIGURATION._type,
  },
  TENANT_HAS_APP_INSTANCE: {
    _type: 'netskope_tenant_has_app_instance',
    sourceType: Entities.TENANT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.APP_INSTANCE._type,
  },
};
