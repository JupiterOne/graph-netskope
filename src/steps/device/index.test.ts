import { executeStepWithDependencies } from '@jupiterone/integration-sdk-testing';
import { buildStepTestConfigForStep } from '../../../test/config';
import { Recording, setupProjectRecording } from '../../../test/recording';
import { IntegrationSteps } from '../constants';
import { omit } from 'lodash';

// See test/README.md for details
let recording: Recording;
afterEach(async () => {
  await recording.stop();
});

test('build-tenant-and-device-relationships', async () => {
  recording = setupProjectRecording({
    directory: __dirname,
    name: 'build-tenant-and-device-relationships',
  });

  const stepConfig = buildStepTestConfigForStep(IntegrationSteps.DEVICES);
  const stepResult = await executeStepWithDependencies(stepConfig);
  expect(stepResult).toMatchStepMetadata(stepConfig);
  expect(omit(stepResult.collectedEntities[0], ['_rawData'])).toMatchSnapshot();
});
