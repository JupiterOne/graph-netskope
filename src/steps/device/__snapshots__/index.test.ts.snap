// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build-tenant-and-device-relationships 1`] = `
Object {
  "_class": Array [
    "Device",
  ],
  "_key": "netskope_device:CLW247-360",
  "_type": "netskope_device",
  "category": "Windows",
  "clientInstallTime": 1628251103,
  "clientVersion": "94.1.1.960",
  "createdOn": undefined,
  "deviceId": "CLW247-360",
  "displayName": "CLW247-360",
  "id": "CLW247-360",
  "lastEvent": "System shutdown",
  "lastEventActor": "System",
  "lastEventNpaStatus": "Steering Disabled",
  "lastEventOccurredOn": 1653479089,
  "lastEventStatus": "Disabled",
  "lastSeenOn": 1653479089,
  "make": "Dell Inc.",
  "model": "Vostro 3580",
  "name": "CLW247-360",
  "os": "Windows",
  "osVersion": "10.0 (2009)",
  "serial": "[REDACTED]",
}
`;
